Run shell sript to set up SilverStripe Behat test session quickly

jeffreyguo opened the pull request:
We have SilverStripe Behat extension and test session, but it's not easy to set up a Behat test session. We need to do lots of settings and configuration after run composer install for these two packages. Here I  created a shell script to do the settings and configurations automatically.
`$ sh vendor/silverstripe/behat-extension/appendixes/SS-behat-quick-setup.sh`
NOTE: suppose "silverstripe/behat-extension" is composer installed.

jeffreyguo commented:
Actually I created a new folder `appendixes` and put all files in it.

tractorcow commented:
Is there markdown documentation for all of this? It's hard to make sense of what file has what relevance.

Are these sample files, or files that will be actually run to create all test sessions going forward?

jeffreyguo commented:
No, there is no markdown documentation for this integration. They are all from my personal experience. I listed all the committed files in a tree structure as below, added my comments. I've also added one step to run the Shell script in README.md.
The sample files will be actually run in creating Behat test session. They are basically fundamentals in Behat test session. I've added warnings in the Shell script to remind users to change the sample files accordingly.

├── appendixes
│   ├── SS-behat-quick-setup.sh (one Shell script to set up Behat test session for all)
│   ├── behat.yml (configure test instance)
│   └── mysite
│       ├── _config
│       │   └── behat.yml (configuration for test state file)
│       ├── code
│       │   └── testing
│       │       ├── FakeDatabase.php (for running fake database)
│       │       ├── FakeObject.php (for creating fake object)
│       │       ├── SampleFakeManager.php (instantiate fake objects)
│       │       ├── SampleFakeManagerRequestFilter.php (Register fake web services)
│       │       └── SampleTestSessionExtension.php (reset test session and the fake database)
│       └── tests
│           ├── behat
│           │   └── features
│           │       ├── bootstrap
│           │       │   └── Context
│           │       │       └── FeatureContext.php (write your Behat step definitions here)
│           │       └── sample.feature (a sample feature file, put your test scenarios here, you can rename it or create another feature files)
│           └── fixtures
│               ├── FakeDatabase.json (configure fake database in a json format)
│               └── SS-sample.sql (db sql for running a test session)

tractorcow commented:
@chillu do you have any thoughts about this? Is this the best way to bootstrap the behat environment? I thought that there was already a task for generating stub feature context, etc.

`vendor/bin/behat --init @mymodule`

Doesn't this simply duplicate this function in a silverstripe specific way?

chillu commented:
First off, thanks for trying to streamline this process Jeffrey!

 * `behat --init` should take care of copying files around and populating templates, since it's the Behat recommend approach. It can also do a better job at documenting accepted CLI parameters, interactively asking for stuff like `base_url`. etc. In my opinion, doing half of the operations in a shell script and the other half via `behat --init` will confuse devs.
 * `FakeObject` and `FakeDatabase` should be added as a composer dependency from https://github.com/chillu/fakedatabase, not inlined here
 * On docs, I've actually tried to write a tutorial for this a while ago on a branch: https://github.com/silverstripe-labs/silverstripe-behat-extension/blob/webservice-tutorial/docs/webservice-mocking.md. It got a bit too complicated to explain all the bits, and I got stuck thinking that this should be more integrated into either the behat-extension or a separate module - too much boilerplate code. Jeffrey, do you want to review these docs and check how up-to-date they still are?
 * There's too much domain-specific stuff in the templates which is hard to identify as placeholders. Instead of `AddressRight` and `Iguana2`, we should have something like `
 * `GatewayCommon` is used in the templates, but not defined anywhere
 * `SampleFakeManagerRequestFilter` and `SampleTestSessionExtension` aren't good names. You'll likely use them without modification, so it should just be `FakeManagerRequestFilter` and `FakeManagerTestSessionExtension`.
 * Given we already have a `testsession` module which is pretty much exclusively used for Behat testing (and some manual regression testing), should we just move `FakeManagerRequestFilter` and `FakeManagerTestSessionExtension` and `_config/behat.yml` there? Arguably, JSON fakes can be used through static files, they don't need to be defined dynamically through Behat steps - so it's a weak dependency.

jeffreyguo commented:
@chillu , thanks for your review and comments. I've done some updates based on your comments and pushed a new commit [2f5794a].

- `behat --init` is moved to the last step, but I need more time to do an interactive prompt for entering `base_url` parameter. Anyway, this could be treated as the initial version of Behat quick setup.
- `FakeObject` and `FakeDatabase` are added from a composer package
- I will have a look the "Mocking SOAP Webservices with Phockito and Behat"
- domain-specific stuff is removed
- GatewayCommon: removed
- Renamed `SampleFakeManagerRequestFilter` and `SampleTestSessionExtension` to `FakeManagerRequestFilter` and `FakeManagerTestSessionExtension` respectively
- You can see here there are 3 modules: SS-Behat-Extension, test-session and fakedatabase. But there are more than the three modules in Behat test session setup. When a beginner runs composer install, the three modules are placed in /vendor/silverstripe/ or project root folder. However, there are some configurations and settings which should be put under /mysite/ directory. E.g.
-- Append two parameters TestSessionEnvironment and TestSessionController to /mysite/_config/config.yml
-- Set up /mysite/code/testing/FakeManagerTestSessionExtension.php
-- Add a test db sql like /mysite/tests/fixtures/SS-sample.sql
......
For details please check the Shell script "/appendixes/SS-behat-quick-setup.sh". 

The purpose of this PR is simple, it is to help a beginner set up a Behat test environment quickly without troubles. Run one command and then Behat test is ready for you.

jeffreyguo commented:
@chillu,  I've added an interactive prompt for entering `base_url` parameter, see commit [813de11].

chillu commented on appendixes/SS-behat-quick-setup.sh:
```diff
@@ -0,0 +1,37 @@
+echo "\n" >> $path/mysite/_config/config.yml
+cat $path/vendor/silverstripe/behat-extension/appendixes/mysite/_config/config.yml >> $path/mysite/_config/config.yml
+cp -Rv $path/vendor/silverstripe/behat-extension/appendixes/mysite/code/testing/ $path/mysite/code/testing/
+cp -Rv $path/vendor/chillu/fakedatabase/src/ $path/mysite/code/testing/
```
It shouldn't be necessary to copy `fakedatabase` files into `mysite`, right? Composer/PHP autoloading should make those files available from their `vendor/` location. If you need customisation of these, subclassing is your friend.

chillu commented on appendixes/mysite/tests/behat/features/bootstrap/Context/FeatureContext.php:
```diff
@@ -0,0 +1,121 @@
+ * Context automatically loaded by Behat.
+ * Uses subcontexts to extend functionality.
+ */
+class FeatureContext extends SilverStripeContext {
```
This file will get overwritten by the `behat --init` in your shell script, see https://github.com/jeffreyguo/silverstripe-behat-extension/blob/master/src/SilverStripe/BehatExtension/Console/Processor/InitProcessor.php. Which is what I was referring to in my earlier comments: FIles should be copied by that `InitProcessor.php`, not a shell script.

chillu commented on composer.json:
```diff
@@ -32,7 +32,8 @@
 		"silverstripe/testsession": "*",
-		"silverstripe/framework": "^4.0.0"
+		"silverstripe/framework": "^4.0.0",
+		"chillu/fakedatabase": "*"
```
If you look at my comment starting with "Given we already have a testsession module...", I recommended to add this dependency and its two custom classes to the `testsession` module instead. You can use `testsession` without `behat-extension`, and the FakeDatabase stuff can be useful without Behat

chillu commented on appendixes/mysite/tests/behat/features/bootstrap/Context/FeatureContext.php:
```diff
@@ -0,0 +1,121 @@
+            parent::getTestSessionState(),
+            array(
+                'useFakeManager' => true,
+                'importDatabasePath' => BASE_PATH .'/mysite/tests/fixtures/SS-sample.sql',
```
You're assuming that every Behat usage will rely on a default SQL file. This adds additional overhead to each scenario execution, since we need to run `dev/build` on top of that SQL import anyway to avoid schema mismatches between the SQL file and the actual schema defined in code. Hiding fixtures away in SQL files is kind of an anti-pattern for testing, since it creates "magic state" that just exists in test executions, rather than explicitly creating fixtures through step definitions. For these two reasons, I think we should remove the SQL file from here, and simply document this as an advanced use case in the README.

jeffreyguo commented:
Thanks @chillu for your feedback.

1. Removed copying `fakedatabase ` package, see new commit [862a786].
2. I tested and confirmed that `behat --init` does not override FeatureContext.php file, I copied content from .../InitProcessor.php to FeatureContext.php file and added test session stuff.
3. We do have `testsession` , `behat-extension` and `fakedatabase` modules. However, they are individual modules. The beginners have no idea on the integration. As I said in above comments, the purpose of this PR is simple, it is to help a beginner set up a SilverStripe Behat test environment quickly without troubles. Run one command and then Behat test is ready for you. I know Behat test can be run without testsession and session database, but I believe it's good to keep them in a quick setup as doing this will let the beginners know it's a good solution to run each test with a temp database as we did and are doing in our projects.

chillu commented on appendixes/mysite/code/testing/FakeManager.php:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @var type
+     */
+    protected $addressRightGateway;
```
Still contains domain specific stuff

tractorcow commented:
My overall feeling for this pull request is that it's going to be a huge code maintenance burden.

jeffreyguo commented:
@chillu , I've checked all files and removed the domain specific stuff, see new commit [cb9fd1b].
@tractorcow , no worries, I put all my updates in a separate folder `appendixes`, they don't impact the existing functionality if users don't touch this folder.

anselmdk commented:
I agree with @tractorcow that this feels like a lot of boilerplate, that should maybe rather go into it's own module?
@jeffreyguo I very much welcome your efforts in trying to streamline setting up a project with Behat. I think this is very necessary. I've myself posted about it to the dev list a few weeks ago, as well as created this short instruction: https://gist.github.com/anselmdk/ee38e089080e3c0b4b88

dhensby commented:
Getting a quick set-up is something that the module should help achieve out of the box, IMO

tractorcow commented:
I don't want to maintain all these scripts that will very quickly become obsolete. =/ Especially if they are db schema specific.

jeffreyguo commented:
It doesn't matter. I've already moved this one into my personal repo https://github.com/jeffreyguo/SS-Behat-quicksetup

anselmdk commented:
@jeffreyguo as mentioned, I think so far it suits better having it in it's own module. However I'd have liked it to be an extension. It seems your module is actually a copy of the entire module?
